### ISSUE TITLE
Please upload a "release" source tarball

### ISSUE THREAD

**Issue by mbakke** [ninja-build/ninja#1728]
Hello,

The source archives generated by GitHub are not deterministic: they may get re-generated in the future with different timestamps, which breaks source-based distributions.

Please consider uploading a copy of the source code alongside the binary releases so that the URL and checksum stays valid "forever".

**Comment by jonesmz** [comment]
Can you elaborate on why github would override previously published source archives?

**Comment by mbakke** [comment]
I assume because they have expired from cache or something along those lines. GitHub support confirmed a few years back that the "archive" downloads are unreliable and may get re-generated at any time.

I tried downloading some old Ninja releases and did get the same checksum as they had originally, so this issue is not extremely prevalent.  Yet it does happen frequently enough (at least for less-popular repositories) that distributions prefer to git clone instead of downloading the tarballs.

**Comment by jhasse** [comment]
Even the ones on a release? IMHO this is a GitHub bug.

**Comment by jonesmz** [comment]
@mbakke Can you confirm that this is specifically for this list of source bundles? https://github.com/ninja-build/ninja/releases

Or is it this one? https://github.com/ninja-build/ninja/archive/master.zip

**Comment by mbakke** [comment]
@jonesmz The latter.  Note how the URL of uploaded files are different:

* Generated on demand (and cached), subject to change: https://github.com/ninja-build/ninja/archive/v1.10.0.zip

* Uploaded file, hosted on S3 behind the scenes, immutable: https://github.com/ninja-build/ninja/releases/download/v1.10.0/ninja-linux.zip

I would like to have a copy of the current source code uploaded alongside the binary releases so they won't ever expire and change checksum.

**Comment by jonesmz** [comment]
I'm pretty sure that's already what happens.

What isn't working correctly for you?

**Comment by mbakke** [comment]
Things are fine now, until GitHub decides to re-generate the cached zipball and tarball.  I just wanted to make you aware that it can happen at any time and break the checksums.

If you don't consider that to be a problem, feel free to close this.

**Comment by jhasse** [comment]
I do consider that to be a problem. But I think this is a GitHub bug, projects shouldn't be required to work-around it by uploading a redundant source tarball.

Did you ask GitHub support about *release* source tarballs in particial?

**Comment by mbakke** [comment]
To my knowledge there is no difference between "release" tarballs, those under the "tags" tab, and arbitrary commits.  I.e. the `/archive/<commit>.tar.gz` URL uses the same machinery in all cases.

That said, it has been a couple of years, so I'll ask GitHub support if things have changed in the mean time.  Thanks for bearing with me on this.  :-)

**Comment by jmillikin** [comment]
Has there been any progress on getting stable source archives for Ninja releases?

To clarify some of the questions in this thread:
* URLs such as https://github.com/ninja-build/ninja/archive/refs/tags/v1.12.0.tar.gz are generated by GitHub on request -- there might be some caching behind the scenes, but fundamentally they're equivalent to wrapping the `git archive` command in an HTTP handler.
* The Git / GitHub archive generator does not guarantee byte-level determinism, for example the gzip compression is allowed to change without warning. This means the checksum of generated source archives might change even if the source files are identical.
* The source files can also change, for example if the `v1.12.0` tag is overwritten, or via certain fields in `.gitattributes`. This has affected other major projects in the past (including Kubernetes).
* GitHub does not plan to guarantee stable checksums for auto-generated archives: https://github.com/orgs/community/discussions/45830

In other words, Ninja does not currently have reliable source releases in the same way that (for example) Meson or CMake do.

The recommended best solution for this is to upload a source archive for the release. It's fine to use the generated one, since that basically just takes whatever GitHub was running that day and pins it.

```
$ curl -L -o ninja-v1.12.0.tar.gz https://github.com/ninja-build/ninja/archive/refs/tags/v1.12.0.tar.gz
$ (upload ninja-v1.12.0.tar.gz as a release artifact)
```

**Comment by jhasse** [comment]
> GitHub does not plan to guarantee stable checksums for auto-generated archives:

I skimmed over the discussion and it looks to me that GitHub *does* guarantee it:

> We're sorry for the breakage, **we're reverting the change,** [...]

I'm using GitHub release tarballs with checksums in other projects for dependencies and never had a problem.

So to sum it up: Use the GitHub generated source tarball: https://github.com/ninja-build/ninja/archive/refs/tags/v1.12.0.tar.gz
If its checksum changes, open an issue.
